website/src/components/TutorialComponents/CodeBlock.tsx

```diff
@@ -120,20 +125,63 @@ function CodeEditor(props: {
     const onClickReset = () => {
         setTmpCode(code.slice()); // use copy not reference
         setCode(props.code);
-        setDisabled(true);
+        // setDisabled(true);
+        setHasFocus(false);
         setAllowUndo(true);
         setTimeout(() => {
-            setDisabled(props.disabled);
+            // setDisabled(props.disabled);
+            setHasFocus(true);
             setAllowUndo(false);
         }, 3000);
     }
 
     const onClickUndo = () => {
-        console.log(tmpCode)
         setCode(tmpCode);
     }
 
-    //   prismIncludeLanguages(Prism);
+    const handleFocus = () => {
+        const selectObj = window.getSelection();
+        if (selectObj.rangeCount === 0) {
+            // when focusing on the editor without using the mouse, 
+            // merely from the Tab key
+            const range = new Range();
+            range.collapse(true);
+            selectObj.addRange(range);
+        }
+        setHasFocus(true);
+    }
+
+    const handleBlur = () => {
+        setHasFocus(false);
+    }
+
+    const handleKeydown = (e: KeyboardEvent) => {
+        const editor = editorRef.current;
+        if (e.key === 'Escape' && !props.disabled) {
+            if (e.target === editor) {
+                e.stopImmediatePropagation();
+                e.stopPropagation();
+                e.preventDefault();
+                editor.focus(); // sometimes blur won't fire without focus first
+                editor.blur();
+            }
+        }
+    }
+
+
+
+    useEffect(() => {
+        document.addEventListener('keydown',
+            handleKeydown,
+            {
+                capture: true,
+            }
+        );
+        return () => {
+            document.removeEventListener('keydown', handleKeydown);
+        };
+    }, []);
```

Comment:
handleKeyDown will depend on "disabled",

- wrap handleKeyDown using useCallback and put `disabled` in the dependency list
- put handleKeyDown in the dependency list of useEffect

Reply:
https://github.com/microsoft/z3guide/pull/109/commits/42853c2a32ed5c7a9649958b661224a3eccfb1fd this?